macrosynergy/pnl/historic_portfolio_volatility.py

```diff
@@ -38,113 +35,156 @@ def expo_weights_arr(lback_periods: int, half_life: int, *args, **kwargs) -> np.
+                x=piv_ret[c_a].values,
+                y=piv_ret[c_b].values,
+                lback_periods=lback_periods,
                 remove_zeros=remove_zeros,
             )
-            cov_matr[iA, iB] = est_vol  # TODO (iA, iB) is same as (iB, iA) 
-            cov_matr[iB, iA] = est_vol  # TODO (iA, iB) is same as (iB, iA)
+            # if either of the return series is not available, set the covariance to NaN
+            if np.isnan(x).all() or np.isnan(y).all():
+                cov_matr[i_a, i_b] = np.nan
+                cov_matr[i_b, i_a] = np.nan
+                continue
+
+            w = weights_func(
+                lback_periods=min(lback_periods, len(x)),
+                half_life=min(half_life, len(x)),
+            )  # inherently fast, and cached anyway
+
+            est_vol = _weighted_covariance(x=x, y=y, w=w)
+            cov_matr[i_a, i_b] = est_vol
+            if i_a != i_b:
+                cov_matr[i_b, i_a] = est_vol
 
     assert np.all((cov_matr.T == cov_matr) ^ np.isnan(cov_matr))
 
     return pd.DataFrame(cov_matr, index=piv_ret.columns, columns=piv_ret.columns)
 
 
+def _calculate_portfolio_volatility(
+    trigger_indices: pd.DatetimeIndex,
+    pivot_returns: pd.DataFrame,
+    pivot_signals: pd.DataFrame,
+    lback_periods: int,
+    nan_tolerance: float,
+    portfolio_return_name: str,
+    *args,
+    **kwargs,
+    # weights_func: Optional[np.ndarray],
+):
+    """
+    Calculate the portfolio volatility for each trigger date.
+    Backed function for `_hist_vol`, to increase readability.
+
+    :param <pd.DatetimeIndex> trigger_indices: the DateTimeIndex of the trigger dates.
+    :param <pd.DataFrame> pivot_returns: the pivot table of the contract returns.
+    :param <pd.DataFrame> pivot_signals: the pivot table of the contract signals.
+    :param <int> lback_periods: the number of periods to use for the lookback period
+        of the volatility-targeting method. Default is 21.
+    :param <int> half_life: Refers to the half-time for "xma" and full lookback period
+        for "ma". Default is 11.
+    :param <callable> weights_func: the function to use for the weights array. Default
+        is None, which means that the weights are equal.
+    :param <float> nan_tolerance: maximum ratio of NaNs to non-NaNs in a lookback window,
+        if exceeded the resulting volatility is set to NaN. Default is 0.25.
+    :param <str> portfolio_return_name: the name of the portfolio return series.
+    :param <dict> kwargs: additional keyword arguments to pass to the variance-covariance
+        estimation function.
+    """
+
+    return_values = []
+
+    for trigger_date in trigger_indices:
+        td = trigger_date
+        lbx = -1 * int(np.ceil(lback_periods * (1 + nan_tolerance)))
+        # account for possible NA drops and create a bigger slice
+        piv_ret = pivot_returns.loc[pivot_returns.index <= td].iloc[lbx:]
+        vcv: pd.DataFrame = _estimate_variance_covariance(
```

Comment:
Maybe add a comment that we are estimating the variance-covariance matrix? Or is this self-evident?